CL-844 Add expected nb of users to user counts / user exports

adessy opened the pull request:
(CL-844)

This PR adds `expected_users` info to the responses of the following endpoints:
```ruby
get 'users_by_gender'
get 'users_by_birthyear'
get 'users_by_education'
get 'users_by_custom_field/:custom_field_id'

get 'users_by_gender_as_xlsx'
get 'users_by_birthyear_as_xlsx'
get 'users_by_education_as_xlsx'
get 'users_by_custom_field_as_xlsx/:custom_field_id'
``` 

The `expected_users` data indicates the number of users that the sample would have contained if it was perfectly representative. It is only available for custom fields with a reference distribution. `expected_users` is added as an extra column in xlsx endpoints and as an extra data series in json endpoints.

The stats_users_controller underwent a complete overhaul. So I'm afraid the diff won't be of much help for the review. Sorry for that. The main goal of the refactoring was to reuse the same underlying logic for all the endpoints (domicile is still an exception). 

There is still a lot to improve in the code, but I tried to leave it in a better state than before.
Happy reviewing 🙂

adessy committed:
CL-844 Remove class variable usage (bad practice)

adessy committed:
CL-844 Extract duplicated code into a method: find_users

adessy committed:
CL-844 Extract duplicated code into a method: count_users

adessy committed:
CL-844 Rework the logic to count the #users by custom field value

adessy committed:
CL-844 Infer the custom field from the request path

Default custom fields are inferred from the request path instead of being hardcoded. This will help us to remove more duplicated code in future refactorings.

Nothing much changes for user-defined custom fields which are identified using the 'custom_field_id' path parameter.

adessy committed:
CL-844 Remove duplicated code

adessy committed:
CL-844 Make a few methods private

adessy committed:
CL-844 Extract xlsx mime type into a constant

adessy committed:
CL-844 Further simplifications to make the code more consistent

adessy committed:
CL-844 Remove some more duplicated code

Now most endpoints for user custom fields stats are calling the same action (`users_by_custom_field`). This introduces some breaking changes because now stats for education and gender also include information about the custom field options. (But this makes the endpoints more consistent.)

Domicile is still a particular case.

adessy committed:
CL-844 Simplify users_by_custom_field

adessy committed:
CL-844 Use the same action to implement most xlsx exports

All xlsx exports for user custom fields, except for domicile, are now implemented using the same controller action. It introduced some minor changes in the exports, but made overall more consistent.

adessy committed:
CL-844 Improve minor things

adessy committed:
CL-844 Fix bug in unknown count in users_by_domicile exports

It was always reporting 0 for the 'unknown' option in users_by_domicile exports.

adessy commented:
Hey @jamesspeake, here's what I've got for now. The refactoring part is mostly done (even it's clearly not perfect), but didn't have the time to move on to adding the expected counts. Lost a bit of time along the way because of some ActiveRecord issue with the `count` method. Sorry the diff is pretty difficult to read because most of the code changed.

Domicile is still a particular case because the way `Area`s are used as options without being linked to any `CustomFieldOption`s.

Next steps are:
- adding the expected counts to the result of `user_counts` if a reference distribution exists for the corresponding custom field.
- find a way to output this as an xlsx (the helper methods provided by `XlsxService` are either lacking in flexibility – `generate_field_stats_xlsx` or `generate_res_stats_xlsx` – or more complex to use than they ought to be – `generate_xlsx`)

adessy committed:
Remove unnecessary condition

adessy committed:
CL-610 Merge branch 'CL-610/representativeness-reference-distribution' into CL-844/add-expected-nb-users-to-stats-users-endpoints

# Conflicts:
#	back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb

adessy committed:
CL-844 Fix rubocop offenses

adessy committed:
CL-844 Refactor total_count method

- better name
- make it private
- use memoization

adessy committed:
CL-844 Add the expected nb of users to the responses of users_by_custom_field

adessy committed:
CL-844 Fix the conversion of keys for expected counts into option keys

adessy committed:
CL-844 Memoize user_counts in StatsUsersController

adessy committed:
CL-844 Fix the number of users used to compute expected counts

In order to be comparable with user counts, expected counts must be computed from the nb of users with a response (for the custom field), not from the total number of users.

I also removed some parameters that were not used and memoized the result of expected counts.

adessy committed:
CL-844 Fix bad formatting

adessy committed:
CL-844 Add expected user count to xlsx exports

adessy committed:
CL-844 Reorder private methods in UserCustomFields::WebApi::V1::StatsUsersController

adessy committed:
CL-844 Adapt sheetname to custom field in xlsx exports

adessy committed:
CL-844 Revert random changes committed by mistake

adessy committed:
CL-844 Adapt test to breaking changes

Responses from web_api/v1/stats/users_by_custom_field endpoint now include the expected nb of users for a representative sample.

adessy committed:
CL-844 Fix option details in responses of .../stats/users_by_... endpoints

adessy committed:
CL-844 Adapt xlsx-export examples to breaking changes

adessy committed:
CL-844 Use json_response_body helper

adessy committed:
CL-844 Use helper method to set up the admin JWT token

adessy committed:
CL-844 Move header specification out of before block

adessy committed:
CL-844 Rework stats_users_spec.rb

- simplify setup (before group)
- reduce code repetition
- fix @timezone that was nil

adessy committed:
CL-844 Test presence of expected counts in responses

Test that endpoints for user counts for select custom field also include expected counts in responses if a reference distribution is available.

adessy committed:
CL-844 Speed up test using TestProf helpers

adessy committed:
CL-844 Handle the case when the regex doesn't match

Add safeguard navigation to handle the cases where the regex does not match. I know the syntax looks awful...

adessy committed:
CL-844 Fix rubocop offenses

adessy committed:
CL-844 Remove unused method

adessy committed:
CL-844 Rework tests for stats/users_by_domicile(_as_xlsx)

Most noteworthy changes:
- Use share examples to test the xlsx version.
- Include a user without domicile as a regression test for commit f172390dab8bc67f5509a90e4602cedb7f1ada3d

adessy committed:
CL-844 Add tests for XlsxService#xlsx_from_columns and XlsxService#xlsx_from_rows

adessy committed:
CL-844 Add some extra tests for RefDistribution model

jamesspeake commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb:
```diff
@@ -4,256 +4,185 @@ module UserCustomFields
+            users.select("custom_field_values->'#{custom_field.key}' as field_value")
+          when 'multiselect'
+            users.joins(<<~SQL.squish).select('cfv.field_value as field_value')
+              LEFT JOIN (
```
Slightly concerned that this query will be slow. Don't know the numbers involved for our biggest datasets.

jamesspeake commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb:
```diff
@@ -4,256 +4,185 @@ module UserCustomFields
+          # Warning: The method +count+ cannot be used here because it introduces a SQL syntax
+          # error while rewriting the SELECT clause. This is because the 'field_value' column is a
+          # computed column and it does not seem to be supported properly by the +count+ method.
+          counts = field_values.order('field_value')
```
Again, be interested to know what this is doing at a database level.

adessy committed:
CL-844 Merge branch 'master' into CL-844/add-expected-nb-users-to-stats-users-endpoints

# Conflicts:
#	back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb
#	back/engines/commercial/user_custom_fields/spec/acceptance/stats_users_spec.rb

jamesspeake reviewed (approved):
OK, I've tried a version of what those SQL queries seem to be doing against the largest user base in production and I'm happy that there isn't actually a performance problem there.

adessy committed:
CL-844 Add reference distribution relationship to serialized user custom fields

adessy committed:
CL-844 Remove experiment that leaked into the code

adessy committed:
CL-844 Fix typos

jamesspeake reviewed (approved):
Approved - As discussed, it just needs a little more clarity in the has_one / belongs_to comment

adessy committed:
CL-844 Improve comment
